### ISSUE TITLE
Function un-inlining

### ISSUE THREAD

**Issue by nshp** [Vector35/binaryninja-api#2185]
This is a far-off dream of a feature request, but it would be very nice if I could select a chunk of HLIL, mark a few expressions as 'variable', and call it an inlined function or macro. Then binja could search for the IL pattern and mark matches, giving an option to simplify/un-inline the inline function call. The first part could be done as a plugin right now, but the un-inlining part would take some core support (or maybe #1606). I can imagine this playing really well with DWARF or PDB imports, where you have the ground truth for macro expansion/inlining, but also for reversing stripped, optimized code.

**Comment by plafosse** [comment]
I agree this is an awesome feature that we've talked about for awhile. It could probably be done with a hack with analysis passes when they exist, but the real solution will take considerably more effort. UI wise you'd want some way to view these un-inlined functions much like how you'd normally view functions. For the sake of data flow and type propagation you'd like to still treat them as inlined.
